### ISSUE TITLE
Using the plugin on elasticsearch-6.2.2, after a couple of hours, then all the search requests got no responses

### ISSUE THREAD

**Issue by li910802** [o19s/elasticsearch-learning-to-rank#153]
We checked the url http://localhost:9214/_nodes/hot_threads?threads=10000&ignore_idle_threads=false, 
and we got the thread stack logs below:

```
0.0% (0s out of 500ms) cpu usage by thread 'elasticsearch[node3][search][T#20]'
 10/10 snapshots sharing following 46 elements
   sun.misc.Unsafe.park(Native Method)
   java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
   java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
   java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
   java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
   java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
   java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
   org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(ReleasableLock.java:55)
   org.elasticsearch.common.cache.Cache.lambda$computeIfAbsent$4(Cache.java:371)
   org.elasticsearch.common.cache.Cache$$Lambda$2714/1659733949.accept(Unknown Source)
   org.elasticsearch.common.cache.Cache$CacheSegment.lambda$get$0(Cache.java:224)
   org.elasticsearch.common.cache.Cache$CacheSegment$$Lambda$2758/1471429601.apply(Unknown Source)
   java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
   java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
   java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2155)
   org.elasticsearch.common.cache.Cache$CacheSegment.get(Cache.java:215)
   org.elasticsearch.common.cache.Cache.get(Cache.java:344)
   org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:370)
   com.o19s.es.ltr.feature.store.index.Caches.cacheLoad(Caches.java:139)
   com.o19s.es.ltr.feature.store.index.Caches.loadFeatureSet(Caches.java:129)
   com.o19s.es.ltr.feature.store.index.CachedFeatureStore.loadSet(CachedFeatureStore.java:51)
   com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:140)
   com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:45)
   org.elasticsearch.index.query.AbstractQueryBuilder.toQuery(AbstractQueryBuilder.java:97)
   org.elasticsearch.index.query.AbstractQueryBuilder.toFilter(AbstractQueryBuilder.java:119)
   org.elasticsearch.index.query.BoolQueryBuilder.addBooleanClauses(BoolQueryBuilder.java:409)
   org.elasticsearch.index.query.BoolQueryBuilder.doToQuery(BoolQueryBuilder.java:382)
   org.elasticsearch.index.query.AbstractQueryBuilder.toQuery(AbstractQueryBuilder.java:97)
   org.elasticsearch.index.query.QueryShardContext.lambda$toQuery$2(QueryShardContext.java:304)
   org.elasticsearch.index.query.QueryShardContext$$Lambda$2740/257320679.apply(Unknown Source)
   org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:316)
   org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:303)
   org.elasticsearch.search.SearchService.parseSource(SearchService.java:707)
   org.elasticsearch.search.SearchService.createContext(SearchService.java:558)
   org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:534)
   org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:330)
   org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:316)
   org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:312)
   org.elasticsearch.search.SearchService$3.doRun(SearchService.java:1002)
   org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:672)
   org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
   org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   java.lang.Thread.run(Thread.java:748)
   
0.0% (0s out of 500ms) cpu usage by thread 'elasticsearch[node3][search][T#27]'
 10/10 snapshots sharing following 46 elements
   sun.misc.Unsafe.park(Native Method)
   java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
   java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
   java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
   java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
   java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
   org.elasticsearch.common.cache.Cache$CacheSegment.remove(Cache.java:290)
   org.elasticsearch.common.cache.Cache.evictEntry(Cache.java:714)
   org.elasticsearch.common.cache.Cache.lambda$computeIfAbsent$4(Cache.java:372)
   org.elasticsearch.common.cache.Cache$$Lambda$2714/1659733949.accept(Unknown Source)
   org.elasticsearch.common.cache.Cache$CacheSegment.lambda$get$0(Cache.java:224)
   org.elasticsearch.common.cache.Cache$CacheSegment$$Lambda$2758/1471429601.apply(Unknown Source)
   java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
   java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
   java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2155)
   org.elasticsearch.common.cache.Cache$CacheSegment.get(Cache.java:215)
   org.elasticsearch.common.cache.Cache.get(Cache.java:344)
   org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:370)
   com.o19s.es.ltr.feature.store.index.Caches.cacheLoad(Caches.java:139)
   com.o19s.es.ltr.feature.store.index.Caches.loadFeatureSet(Caches.java:129)
   com.o19s.es.ltr.feature.store.index.CachedFeatureStore.loadSet(CachedFeatureStore.java:51)
   com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:140)
   com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:45)
   org.elasticsearch.index.query.AbstractQueryBuilder.toQuery(AbstractQueryBuilder.java:97)
   org.elasticsearch.index.query.AbstractQueryBuilder.toFilter(AbstractQueryBuilder.java:119)
   org.elasticsearch.index.query.BoolQueryBuilder.addBooleanClauses(BoolQueryBuilder.java:409)
   org.elasticsearch.index.query.BoolQueryBuilder.doToQuery(BoolQueryBuilder.java:382)
   org.elasticsearch.index.query.AbstractQueryBuilder.toQuery(AbstractQueryBuilder.java:97)
   org.elasticsearch.index.query.QueryShardContext.lambda$toQuery$2(QueryShardContext.java:304)
   org.elasticsearch.index.query.QueryShardContext$$Lambda$2740/257320679.apply(Unknown Source)
   org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:316)
   org.elasticsearch.index.query.QueryShardContext.toQuery(QueryShardContext.java:303)
   org.elasticsearch.search.SearchService.parseSource(SearchService.java:707)
   org.elasticsearch.search.SearchService.createContext(SearchService.java:558)
   org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:534)
   org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:330)
   org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:316)
   org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:312)
   org.elasticsearch.search.SearchService$3.doRun(SearchService.java:1002)
   org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:672)
   org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
   org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   java.lang.Thread.run(Thread.java:748)
```

**Comment by nomoa** [comment]
Thanks for the report,
Is it easy for you to reproduce the issue?
This kind of problem may be hard to reproduce so would it possible for you to give us more information?
What was the search request you sent? I see that it tries to load a featureset were you logging feature vectors?
How many features are in this feature set?
Was it a single node that started to hang or all the nodes in your cluster?
What was the state of the machine, was it struggling on CPU/mem? Could you check gc logs I think elasticsearch enables them by default.
If it's easy to reproduce does it change something if you increase mem cache to 100m adding `ltr.caches.max_mem: 100mb` in your elasticsearch.yml config file.

Thanks for your help.

**Comment by GerardBouchar** [comment]
Hello,
I think i am experiencing the same problem on ElasticSearch V 5.6.7 : 
- The problem occurs after a few search and training (maybe 30 or 50). 
- I can see using the /_cat/thread_pool end point that on some nodes (i have never seen more than two) the number of active search threads increase when i perform search with the ltr plugin  but never decrease.
- there are about 15 features in my feature set.
- most of the time the machine (cpu/mem) seems ok, maybe the heap is a bit high on the node that are stuck.
- i have no choice to restart es on the concerned nodes to be able to use the plugin again.
- While searches with the plugin are stuck, my es cluster is handling all kind of operations in "background" (indexation, search, bulk...) and everything works well.
- when i launch POST /_ltr/_clearcache i see that the stats on each nodes are 0 
`"stats": {
        "total": {
          "ram": 0,
          "count": 0
        },
        "features": {
          "ram": 0,
          "count": 0
        },
        "featuresets": {
          "ram": 0,
          "count": 0
        },
        "models": {
          "ram": 0,
          "count": 0
        }
      }
    }`
but on the nodes with pending ltr search, i see : 
`"stats": {
        "total": {
          "ram": 29752,
          "count": 2
        },
        "features": {
          "ram": 0,
          "count": 0
        },
        "featuresets": {
          "ram": 0,
          "count": 0
        },
        "models": {
          "ram": 29752,
          "count": 2
        }
      }
    }`

bellow is the stack trace i get when i call /_nodes/hot_threads?threads=10000&ignore_idle_threads=false
`
sun.misc.Unsafe.park(Native Method)
       java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
       java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1685)
       java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3320)
       java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1721)
       java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1887)
       org.elasticsearch.common.cache.Cache$CacheSegment.remove(Cache.java:290)
       org.elasticsearch.common.cache.Cache.evictEntry(Cache.java:714)
       org.elasticsearch.common.cache.Cache.lambda$computeIfAbsent$4(Cache.java:372)
       org.elasticsearch.common.cache.Cache$$Lambda$1970/816286349.accept(Unknown Source)
       org.elasticsearch.common.cache.Cache$CacheSegment.lambda$get$0(Cache.java:224)
       org.elasticsearch.common.cache.Cache$CacheSegment$$Lambda$1887/1650549232.apply(Unknown Source)
       java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
       java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
       java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2147)
       org.elasticsearch.common.cache.Cache$CacheSegment.get(Cache.java:215)
       org.elasticsearch.common.cache.Cache.get(Cache.java:344)
       org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:370)
       com.o19s.es.ltr.feature.store.index.Caches.cacheLoad(Caches.java:139)
       com.o19s.es.ltr.feature.store.index.Caches.loadModel(Caches.java:133)
       com.o19s.es.ltr.feature.store.index.CachedFeatureStore.loadModel(CachedFeatureStore.java:56)
       com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:138)
       com.o19s.es.ltr.query.StoredLtrQueryBuilder.doToQuery(StoredLtrQueryBuilder.java:46)
       org.elasticsearch.index.query.AbstractQueryBuilder.toQuery(AbstractQueryBuilder.java:96)
       org.elasticsearch.search.rescore.QueryRescorerBuilder.build(QueryRescorerBuilder.java:172)
       org.elasticsearch.search.SearchService.parseSource(SearchService.java:672)
       org.elasticsearch.search.SearchService.createContext(SearchService.java:485)
       org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:461)
       org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:257)
       org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:343)
       org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:340)
       org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69)
       org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1556)
       org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638)
       org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
       java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
       java.lang.Thread.run(Thread.java:745)`

Thank you for your help.

**Comment by nomoa** [comment]
I found one possible cause this could happen.
If an unexpected exception is thrown while loading the value.
An unexpected exception is very vague but OutOfMemory error is one of them.
For OOM one way to assert this would be to start elasticsearch using the `-XX:+ExitOnOutOfMemoryError` or `-XX:+CrashOnOutOfMemoryError` so that elasticsearch would simple stop.

For other errors the only way would be to inspect all elastic response outputs to see if one or more shards failed to execute the query, there should a trace in the elasticsearch response. Having this information would greatly help to debug this issue.

Thanks!

**Comment by GerardBouchar** [comment]
Thank you ,

What do you mean by  "all elastic response outputs" ?

**Comment by nomoa** [comment]
@GerardBouchar I mean when you read the response elastic returns when searching using the sltr query it's possible that one of them fails and all subsequent calls will trigger the deadlock and hang forever.
If one unexpected error is thrown while the cache is being loaded it should appear in the response you receive from elastic.
It may be a HTTP error code 400 or 500 or a shard failure.
Shard failures are encoded in the response:

```json
{
  "took" : 63,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : 1000,
    "max_score" : null,
    "hits" : [ {
    }]
 }
}
``` 
Here if `failed` is non zero you should have an additional block in the response with all distinct shard failures and usually containing the name of the exception thrown.

Thanks.

**Comment by GerardBouchar** [comment]
Thank you for this precision.

unfortunately, i've got no response at all. A post request using my model just hang up forever.
Also, i am going to check twice but i see nothing in my ElasticSearch logs.

Thank you.

**Comment by nomoa** [comment]
I meant to check the responses of the few requests that succeeded to return something before the ones that hang.
If the first request hangs then it's troublesome but it still could happen in some conditions such as 2 shards of the same index on the same nodes or if you use the sltr query multiple times in a single request.

Basically my assumption is that the bug behavior:
- 1 failure
- all other calls hang

If I could confirm that there is a a failure before it hangs. Inspecting the type of the exception that was thrown would allow me to confirm that we are being hit by the bug I reported upstream.

**Comment by GerardBouchar** [comment]
ok i see. 
I will try to get you such a response.

Thank you.

**Comment by jasontedor** [comment]
@GerardBouchar Also do you have any log entries that contain `uncaught exception in thread`?

**Comment by GerardBouchar** [comment]
Hello Jasontedor
Thank you for your interest.
i do not see anything like this at info log level on the node that seems to block.

**Comment by li910802** [comment]
Sorry to reply late.
The problem reproduced easily. Only restarting all the nodes can solve the problem, and then it came up again.
There are 4 nodes in our cluster.
The search request we sent did rescoring and logging the feature (only 1 feature) at the same time. The problem also came up when we only use the plugin to rescore. 
And we set ltr.caches.max_mem to 100mb or bigger, but it dosen't help.
Now we are debugging the code, if we find the reason, we'll inform you immediately.

**Comment by nomoa** [comment]
@li910802 @GerardBouchar when it's happening again could capture the full stack of the java virtual machine using `jstack` directly on the elastic servers.
Thanks.

**Comment by li910802** [comment]
Hi, @nomoa, we capture the full stack of elasticsearch, but the log file is too big. 
How can I give it to you ? By email ?

**Comment by nomoa** [comment]
@li910802 sure, I've sent you an email to your address @163.com

**Comment by nomoa** [comment]
@li910802 @GerardBouchar there seems to be a deadlock condition when time based eviction happens.

It's possible to disable such eviction process by adding:
```
ltr.caches.expire_after_write: 0
ltr.caches.expire_after_read: 0
ltr.caches.max_mem: 100mb
```

`ltr.caches.max_mem` should ensure that size based eviction does not happen too often because I suppose the deadlock can also occur on such eviction based on size.

Would it be possible for you to test with these settings and let me know if you run into the same deadlock issues (capturing the full stack with jstack will greatly if it happens with these settings)?

Thanks!

**Comment by li910802** [comment]
@nomoa we analyze the org\elasticsearch\common\cache\Cache.java 
and we guess that when two thread happened to delete the same expired entry at the same time, if one thread already delete the expired entry, then the other thread would run into the deadlock.
We changed the code by adding the judgement whether the entry is deleted or not:
```
public V computeIfAbsent(K key, CacheLoader<K, V> loader) throws ExecutionException {
        long now = now();
        // we have to eagerly evict expired entries or our putIfAbsent call below will fail
        V value = get(key, now, e -> {
            try (ReleasableLock ignored = lruLock.acquire()) {
                **if (e.state != State.DELETED) {**
                    evictEntry(e);
                }
            }
        });
	……
}
```
We are running the stress testing, the deadlock dosen't happen by now. 
I'm not sure whether our analysis is right.

**Comment by nomoa** [comment]
@li910802 thanks!

I think we should follow-up on a bug report directly on the elastic repository. My analysis was somewhat similar. For me the problem is that the CompletableFuture internal lock and the lrulock are locked in different orders leading to the deadlock. The conditions for this to happen are still unclear to me but they involve eviction as you mentioned.
I can reproduce this with a stress test locally.

The thread holding the lrulock but waiting on the completable future:
```
"pool-1-thread-5" #18 prio=5 os_prio=0 tid=0x00007f12bc133800 nid=0xc5a waiting on condition [0x00007f1318d23000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000076e4948c8> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at org.elasticsearch.common.cache.Cache$CacheSegment.remove(Cache.java:290)
	at org.elasticsearch.common.cache.Cache.evictEntry(Cache.java:714)
	at org.elasticsearch.common.cache.Cache.lambda$computeIfAbsent$4(Cache.java:372)
	at org.elasticsearch.common.cache.Cache$$Lambda$15/11902308.accept(Unknown Source)
	at org.elasticsearch.common.cache.Cache$CacheSegment.lambda$get$0(Cache.java:224)
	at org.elasticsearch.common.cache.Cache$CacheSegment$$Lambda$28/1616296000.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:422)
	at com.o19s.es.ltr.CacheTest.lambda$test$1(CacheTest.java:21)
	at com.o19s.es.ltr.CacheTest$$Lambda$7/1216410991.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

The thread trying to complete the future (but locked on lrulock before calling `promote()`):
```
"pool-1-thread-2" #15 prio=5 os_prio=0 tid=0x00007f12bc12e800 nid=0xc57 waiting on condition [0x00007f1319027000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000076e146f80> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
	at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
	at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
	at org.elasticsearch.common.util.concurrent.ReleasableLock.acquire(ReleasableLock.java:55)
	at org.elasticsearch.common.cache.Cache.lambda$computeIfAbsent$5(Cache.java:391)
	at org.elasticsearch.common.cache.Cache$$Lambda$26/534170884.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:422)
	at com.o19s.es.ltr.CacheTest.lambda$test$1(CacheTest.java:21)
	at com.o19s.es.ltr.CacheTest$$Lambda$7/1216410991.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

My feeling is that you are right but the race condition could still happen in the sense that the call `segment.delete(key)` is on the key. There are no guarantee that the entry that was detected to be evicted is now the same:
- thread1 calls computeIfAbsent and and detects that the entry needs to be evicted 
- thread2 does the same and actually evicts the entry and then creates a new CompletableFuture for this same key
- thread1 lock the lrulock and calls evictEntry using the **key**
- thread1 finds the newly created CompletableFuture by thread2 and start to wait for its completion. It actually wants to evict the entry that was already evicted by thread2.
- thread2 is locked since it needs lrulock to actually complete the CompletableFuture that thread1 is waiting for. (This is where I find CompletableFuture misleading, I would have guessed that the future is marked done before it starts calling the handles chain).

To sum-up I believe that the issue is because eviction works by passing a key. The eviction detection (isExpired) and actual eviction should happen on the entry not the same key (the entry for this key may have changed in between). 

@jasontedor (sorry to bother you again about the Cache), do you think this deserves following-up with a issue on the elastic repo?
I think this plugin is affected by this issue because there are not so many keys to cache so chances to have contention on the same key are high.

@li910802 concerning the plugin, I'm currently refactoring how the cache works by making it asynchronous. Having a blocking call at query time is not desirable and I will use the asynchronous support that the QueryBuilder offers during rewrite. As a workaround I still think that disabling time based eviction can greatly reduce the chances to enter this deadlock.

**Comment by nomoa** [comment]
For reference I think this was introduced in https://github.com/elastic/elasticsearch/pull/26516.
This would explain why our cluster (running elastic 5.5.2) is not affected, because this fix was backported in 5.5.3.

**Comment by jasontedor** [comment]
Please see elastic/elasticsearch#30461 which I think will address the issues that you're seeing here. This will be included in 5.6.10 and the 6.3.0 releases of Elasticsearch.

**Comment by nomoa** [comment]
@jasontedor many thanks for the fix!
